File: lab4.cs
using System;
using System.Threading;

namespace Threads
{
    public class Token
    {
        public string data;
        public int recipient;
    }

    public class ThreadRunner
    {
        private Token token;

        public void Run(Token token, int n)
        {
            this.token = token;
            for (int i = 1; i <= n; i++)
            {
                Thread thread = new Thread(Func);
                thread.Start(i);
            }
        }

        private void Func(object data)
        {
            lock (token)
            {
                int threadNum = (int)data;
                Console.WriteLine("thread #{0}", threadNum);
                if (token.recipient == threadNum)
                    Console.WriteLine("{0} received", token.data);
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            int n = Convert.ToInt32(Console.ReadLine());
            Token token = new Token();
            token.data = "data";
            token.recipient = n;
            ThreadRunner runner = new ThreadRunner();
            runner.Run(token, n);
        }
    }
}

File: lab4.go
package main

import (
	"fmt"
	"time"
)

type Token struct {
	data string
	recipient int
}

func initial(ch chan Token, token Token) {
	fmt.Println("initial thread")
	ch <- token
}

func thread(ch chan Token, num int) {
	fmt.Println("thread ", num) 
	token := <-ch
	if token.recipient == num {
		fmt.Println(token.data, "received")
	} else {
		ch <- token
	}
}

func main() {
	var n int
	fmt.Scanln(&n)
	token := Token {"some data", n}
	ch := make(chan Token)
	go initial(ch, token)
	for i := 1; i <= n; i++ {
		go thread(ch, i)
	}
	time.Sleep(time.Second)
}